source/agora/node/Registry.d

```diff
@@ -677,8 +698,145 @@ private struct ZoneData
     /// Whether we are authoritative or we're just caching
     public bool authoritative;
 
-    /// The content of the zone
-    public TypedPayload[PublicKey] map;
+    /// Database to store data
+    private ManagedDatabase db;
+
+    /***************************************************************************
+
+         Params:
+           cache_db = Database instance
+
+    ***************************************************************************/
+
+    public this (ManagedDatabase cache_db)
+    {
+        this.db = cache_db;
+
+        this.db.execute("CREATE TABLE IF NOT EXISTS registry_"~ TemplateTableName ~ "_signature " ~
+            "(pubkey TEXT, signature TEXT NOT NULL, sequence INTEGER NOT NULL, " ~
+            "PRIMARY KEY(pubkey))");
+
+        this.db.execute("CREATE TABLE IF NOT EXISTS registry_"~ TemplateTableName ~ "_addresses " ~
+            "(pubkey TEXT, address TEXT NOT NULL, type INTEGER NOT NULL, " ~
+            "FOREIGN KEY(pubkey) REFERENCES registry_"~ TemplateTableName ~ "_signature(pubkey) ON DELETE CASCADE, " ~
+            "PRIMARY KEY(pubkey, address))");
+    }
+
+    /***************************************************************************
+
+         Returns:
+           Total number of name registries
+
+    ***************************************************************************/
+
+    public ulong count () @trusted
+    {
+        auto results = this.db.execute("SELECT COUNT(*) FROM registry_"~ TemplateTableName ~ "_signature");
+        if (!results.empty)
+        {
+            return results.oneValue!(ulong);
+        }
+
+        return 0;
+    }
+
+    /***************************************************************************
+
+         Returns:
+           All registries in persistent storage
+
+    ***************************************************************************/
+
+    public ResultRange getRegistries () @trusted
+    {            
+        return this.db.execute("SELECT pubkey FROM registry_"~ TemplateTableName ~ "_signature");
+    }
+
+    /***************************************************************************
+
+         Get payload data from persistent storage
+
+         Params:
+           public_key = the public key that was used to register
+                         the network addresses
+
+        Returns:
+            TypedPayload of name registry associated with `public_key`
+
+    ***************************************************************************/
+
+    public TypedPayload getRegistryTypedPayload (PublicKey public_key) @trusted
```

Comment:
Regarding naming:

I would recommend to try avoiding repeating the return type name in the method name, and focus on the semantic meaning of the function (This is a soft rule, to consider with good judgement, because for some simple functions it *does* make sense to repeat the type name).

Here you have a function that takes a `PublicKey` and returns a `TypedPayload`. Obviously there's a relationship between the parameter and the return, so you could go with a generic name (`lookup` ? `get` like AAs ? After all it is the main function for this map-like object), or something more domain-specific (e.g. `findRegistered`).

The last option would be to mimic built-in property. For example, for an associative array, one can do `value in myAA` and get a pointer. If you wanted to do this, you can implement the operator overload. However this would require to handle the lifetime of the return value, so it's probably more trouble than it's worth.